pkg/apps/deployment/status.go

```diff
@@ -0,0 +1,52 @@
+package deployment
+
+import (
+	"fmt"
+	"strings"
+
+	appsv1 "k8s.io/api/apps/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	corelistersv1 "k8s.io/client-go/listers/core/v1"
+)
+
+// PodContainersStatus return detailed information about deployment pods and the containers status in human readable format.
+// This can be used for cluster operator condition messages or logging.
+func PodContainersStatus(deployment *appsv1.Deployment, podClient corelistersv1.PodLister) ([]string, error) {
+	deploymentPods, err := podClient.Pods(deployment.Namespace).List(labels.SelectorFromSet(deployment.Spec.Template.Labels))
+	if err != nil {
+		return nil, err
+	}
+	deploymentPodsStates := []string{}
+	for i := range deploymentPods {
+		containerStates := []string{}
+		for _, c := range append(deploymentPods[i].Status.ContainerStatuses, deploymentPods[i].Status.InitContainerStatuses...) {
+			ready := "ready"
+			if !c.Ready {
+				ready = "NOT ready"
+			}
+			switch {
+			case c.State.Running != nil:
+				containerStates = append(containerStates, fmt.Sprintf("container %q is %s running since %s", c.Name, ready, c.State.Running.StartedAt))
+			case c.State.Waiting != nil:
+				containerStates = append(containerStates, fmt.Sprintf("container %q is %s waiting: %s - %s", c.Name, ready, c.State.Waiting.Reason, c.State.Waiting.Message))
```

Comment:
don't use colon `:` in these messages. It's hard to read the final string for a pod